### DISCUSSION TITLE
有氧化 yutto 的计划吗？

### DISCUSSION THREAD

**Discussion by ifsheldon** [yutto-dev/yutto#334] [General]
Hello! 我看见 biliass 好像开始用 Rust 替换部分代码了，想问一下 yutto 本身有氧化的计划吗？我觉得 yutto 非常好用，就是安装有点麻烦，Python 依赖难解，Rust化之后就可以编译成纯可执行文件了；我之前下载视频也稍微看了一下 yutto 的逻辑，用 Rust 实现了我常用一个下载功能。如果有计划的话，可以一起慢慢把里面某一些部分换成 Rust binding

**Comment by SigureMo** [comment]
:joy: yutto 用 rust 重写会比较麻烦些，目前没有明确的计划，biliass 是限于性能在几年前就有用 rust 重写的计划了，对于 yutto 而言，重写之后可能性能提升不大，但个人还是有兴趣的

这边个人主要是担心没有充足的人力和精力投入，而且目前对 rust 了解有限，不过如果有兴趣一起搞的话，倒是可以试试～

> 就是安装有点麻烦，Python 依赖难解

这里问一下，一般是使用什么安装的呢？如果使用自带独立虚拟环境的 pipx、uv，是不需要考虑依赖冲突问题的，Homebrew、docker 也不必说也是独立环境，一般不建议直接使用 pip 在全局环境安装

**Reply by ifsheldon** [reply]
>  这里问一下，一般是使用什么安装的呢？

我是用 homebrew 安装的，安装倒是没什么问题，只是觉得 Python 的依赖太多了，一个是需要包管理器，不管是 homebrew 还是 uv 之类的，需要安装太多辅助的工具；另一个也会占用太多空间。如果可以用静态语言编译的话，那对于用户来说就是下载一个可执行文件就可以了。

另外，我稍微看了一下整个代码库，对于我来说，可能比较复杂的是各种检查和 parse 的逻辑，其余可能还好；我不知道你觉得最麻烦或者最难的地方在哪，可以告诉我一下。

我看了 async Rust 和 asyncio 交互的东西，比我想象的复杂，我感觉用 Rust 重写会更方便（当然不是要求你重写 orz），而不是换成 Rust binding （链接代码会很乱），看看你有没有建议？

**Reply by SigureMo** [reply]
> 一个是需要包管理器，不管是 homebrew 还是 uv 之类的，需要安装太多辅助的工具

还好吧……而且 yutto 基本也没啥依赖

> 另一个也会占用太多空间

除去 Python 解释器没更多空间了吧？代码基本不占空间

这两个理由不足以说明 rust 重写的好处哦，虽然我是比较喜欢使用 rust 重写，但是并不鼓励无脑用 rust 重写

> 我不知道你觉得最麻烦或者最难的地方在哪

rust 重写的话下载管理肯定会麻烦一些，但可能也还好，我没说很麻烦，只是相对于 biliass 而言，yutto 是一个端到端的模块，单测系统并不完善，这就导致了重构过程很难保证不造成一些回归问题

> 我感觉用 Rust 重写会更方便

之前在替换 aiohttp 之前就考虑过类似的事情，想过用 rust binding 替换 Fetcher，但确实交互会麻烦些

当然，如果可以处理交互问题的话，从 Fetcher 开始重写是比较合适的，毕竟是可插拔的，之前已经从 aiohttp 替换成 httpx（#218），成本很低

**Reply by aqua256** [reply]
我最近发现从uv安装yutto, 需要先为biliass装cargo工具链，直接`uv tool install git+https://github.com/yutto-dev/yutto.git@main`会报error，是不是在readme里先给这种依赖加上安装的链接呢，另外我不确定除uv外其他的安装方式 pipx, pip报不报error。

**Reply by SigureMo** [reply]
> 我最近发现从uv安装yutto, 需要先为biliass装cargo工具链，直接`uv tool install git+https://github.com/yutto-dev/yutto.git@main`会报error

嗯嗯，这个我最近会更新下
